System colour

elduddz opened the pull request:
- setting colours as property ADChange Object
- moving description of system colours to after ADChanges call
  - as this hydrates the system colour property

elduddz committed:
setting systemColours

- setting up colours in system start
- using colours in Object (passed as property on ADChanges object)
- Setting colours on Traverse-Group tests to pass Write-Colour calls

ld-at-work committed:
moving SystemColour into ADChange object

MarcDenman commented on UKHO.ADM.Tests/functions/Update-AD.Tests.ps1:
```diff
@@ -2,7 +2,15 @@ InModuleScope $mut {
         Context "with valid cd" {
             Mock Generate-ConfigurationObject       
-            Mock Get-ADChanges { return @{ "CreatedOUs" = 1}}
+            Mock Get-ADChanges { return @{ "CreatedOUs" = 1; SystemColours = @{
```
Again, just double checking. Is this just to make the test run? You have added the SystemColours object but you don't then add any new assertions anywhere.

elduddz replied:
if I didnt add the SystemColours to the Get-ADChanges return, the Update-AD function fails because it uses those values to return the legend of the colours before the run.

MarcDenman commented on UKHO.ADM.Tests/internal/Traverse-Group.Tests.ps1:
```diff
@@ -155,6 +155,7 @@ InModuleScope $mut {
             $group.UserAccountMembers += [ADUserAccount]::new("user1", $domain)
 
             $ADChanges = [ADChanges]::new()
+            $ADChanges.SystemColours = $systemColours
```
Is `$systemColours` null here? What does this line do?

elduddz replied:
ah wait, I missed that one

MarcDenman commented on UKHO.ADM/Functions/Update-AD.ps1:
```diff
@@ -4,31 +4,31 @@ function Update-AD {
         $out = Generate-ConfigurationObject -ConfigData $cd 
 
-        Write-Colour -LinesBefore 1 "UKHO.ADM will perform the following changes:" -Color Blue
+        $ADChanges = Get-ADChanges -ADDomains $out
```
This is now in a different place, above where it was before. Was this on purpose?

It will now output 

- Output of all the changes
- Description of the symbols
- A line saying - "UKHO.ADM will perform the following changes:"
- Change summary

It might be better this way around(save having to scroll to the top maybe to see the description) but dunno. Just checking if on purpose(not mentioned in PR description)

elduddz replied:
All of the Write-Colours lines are now after the Get-ADChanges call as this instanciates the SystemColour properties.

elduddz replied:
alternative might be to instantiate ADChanges before passing into Get-ADChanges

Or

I could put the legend in the Get-ADChanges function

elduddz replied:
Moved the Write-Colour call for the legend types into the creation point of the ADChanges object

Moved all the other "process" points to other areas in the code.

MarcDenman commented on UKHO.ADM.Tests/functions/Update-AD.Tests.ps1:
```diff
@@ -2,7 +2,15 @@ InModuleScope $mut {
+                "remove" = "Red";
+                "modify" = "Yellow";
+                "error" = "Magenta";
+                "header" = "Blue";
```
`header` isn't a great name. Alert? I wouldn't be sure what to use header for and would probably avoid using it.

elduddz replied:
`section`? not sure it is an alert, it is a bit saying I'm Going to do this...

elduddz replied:
`start`
`process`

elduddz replied:
`progress`
`stage`
`phase`

elduddz replied:
header is now `process`

MarcDenman commented on UKHO.ADM/Objects/Objects.ps1:
```diff
@@ -104,12 +104,21 @@ class ADDomain {
     [string]$DistinguishedName
     [ADOrganisationalUnit[]]$OrganisationalUnits
-    [pscredential]$Credential
+    [psCredential]$Credential
```
Just double checking, will this change in capitalisation break anything?

elduddz replied:
not sure where that came in, maybe the autoformating?

elduddz replied:
To answer the question, it won't

MarcDenman replied:
Ok then :)

MarcDenman commented on UKHO.ADM/Objects/Objects.ps1:
```diff
@@ -267,8 +277,8 @@ class ADChanges {
-                Write-Colour "x Failed to add user $($user.UPN) to group $($group.Name) on $($group.Domain.DistinguishedName):" -Color Magenta
-                Write-Colour -StartTab 1 "$outputString" -Color Magenta
+                Write-Colour "x Failed to add user $($user.UPN) to group $($group.Name) on $($group.Domain.DistinguishedName):" -Color $this.SystemColours.error
+                Write-Colour -StartTab 1 "$outputString" -Color $this.SystemColours.error
```
Slight inconsistency I think (not 100% sure. A little bit hazy, use your best judgement). Should `SystemColours.info` be used instead of `SystemColours.error` for this line?

```powershell
Write-Colour "x Failed to add user $($user.UPN) to group $($group.Name) on $($group.Domain.DistinguishedName):" -Color $this.SystemColours.error
Write-Colour -StartTab 1 "$outputString" -Color $this.SystemColours.error
```

Whereas on ln269(and others)

```powershell
Write-Colour "x Failed to add user $($user.SamAccountName) to group $($group.DistinguishedName):" -Color $this.SystemColours.error
Write-Colour -StartTab 1 "$_" -Color $this.SystemColours.info #Echos out the exceptions message
```

elduddz replied:
yes, I think I did a find and replace, missed that double up

elduddz replied:
fixed in latest

elduddz committed:
output of message should be info

elduddz committed:
Removed set on SystemColours

elduddz committed:
removed all text written out in update-AD

elduddz committed:
writing colours at points they are needed

I think this works better as the location is more known.

but I will need to see if objects are affected by this choice.

elduddz committed:
versioned up to 5.1.4

implementation of SystemColour on ADChanges Object
